### PR TITLE
Viewmodel bindings with UI views

### PR DESCRIPTION
**Pull request by vexonius**
Extension methods and helpers are mostly taken from five codestyle repo

https://user-images.githubusercontent.com/28738980/179476914-c5ec15d4-8bfa-46aa-8ba4-b51f0352ee69.mp4

### CODE REVIEW

File: QuizApp/Presentation/Common/Interfaces/BaseViewController.swift [hunk 1]
```diff
@@ -1,8 +1,11 @@
 import Foundation
 import UIKit
+import Combine
 
 class BaseViewController: UIViewController {
 
+    var cancellables: [AnyCancellable] = []
```
Comment (istaudacher) [inline QuizApp/Presentation/Common/Interfaces/BaseViewController.swift]: Please do not add cancellables in BaseViewController. Instead, init it in each VC. Also, use Set<AnyC>

File: QuizApp/Presentation/Common/UIComponents/RoundedTextInput.swift [hunk 2]
```diff
@@ -102,14 +102,15 @@ class RoundedTextInput: UITextField {
 
     private func setBorder() {
         layer.borderWidth = self.isEditing ?
-        DesignConstants.InputComponents.borderWidth.cgFloat :
-        DesignConstants.InputComponents.noBorder.cgFloat
+            DesignConstants.InputComponents.borderWidth.cgFloat :
```
Comment (istaudacher) [inline QuizApp/Presentation/Common/UIComponents/RoundedTextInput.swift:105]: is this indented properly?

File: QuizApp/Presentation/Login/LoginViewController.swift [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import UIKit
 import SnapKit
+import Combine
```
Comment (istaudacher) [inline QuizApp/Presentation/Login/LoginViewController.swift]: this should be first import

File: QuizApp/Presentation/Login/LoginViewController.swift [hunk 4]
```diff
@@ -142,6 +160,62 @@ extension LoginViewController: ConstructViewsProtocol {
...
+            }
+            .store(in: &cancellables)
+
+        viewModel
+            .$isPasswordHidden
+            .assign(to: \.isSecureTextEntry, on: passwordInput)
+            .store(in: &cancellables)
+
+        loginButton
+            .gesture(.tap())
```
Comment (istaudacher) [inline QuizApp/Presentation/Login/LoginViewController.swift:190]: `.tap`

File: QuizApp/Presentation/Login/LoginViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import Foundation
+import Combine
+
+class LoginViewModel {
+
+    @Published var email: String = ""
```
Comment (istaudacher) [inline QuizApp/Presentation/Login/LoginViewModel.swift]: Remove @Published from email & password and make them private.

File: QuizApp/Presentation/Login/LoginViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    @Published var password: String = ""
+    @Published var isLoginButtonEnabled: Bool = true
```
Comment (istaudacher) [inline QuizApp/Presentation/Login/LoginViewModel.swift]: This should be false by default.

File: QuizApp/Presentation/Login/LoginViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    private func validateInputs() {
+        guard
+            !email.isEmpty,
+            !password.isEmpty
+        else {
+            return isLoginButtonEnabled = false
+        }
+
+        isLoginButtonEnabled = true
```
Comment (istaudacher) [inline QuizApp/Presentation/Login/LoginViewModel.swift]: isLoginButtonEnabled = !email.isEmpty && !password.isEmpty

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/Presentation/Common/Interfaces/BaseViewController.swift [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import Foundation
 import UIKit
+import Combine
```
Comment (istaudacher) [inline QuizApp/Presentation/Common/Interfaces/BaseViewController.swift]: do you need this import?

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/Presentation/Common/Interfaces/BaseViewController.swift (hunk 8 above)
Comment (vexonius) [reply QuizApp/Presentation/Common/Interfaces/BaseViewController.swift]: nope

**Review by istaudacher — APPROVED** [review]

**Pull request merged**